### DISCUSSION TITLE
What is the valid regex or the valid characters for the name of a type in a model?

### DISCUSSION THREAD

**Discussion by bubianchi-criteo** [openfga/community#302] [Q&A] [answered]
In a model in DSL format or Json format, it looks like defining types with names that contain `'/'` or `'.'` characters are supported. For example: 
`type test/resource` 
`type test.user`
It works correctly in a openfga server.
Nevertheless, in its playground or in the VSCode extension, the syntax highlighting and parser report them as errors. 
I didn't find in the documentation about a regex or valid characters for `type` names (as well as for ids). 
Is there some doc specifying what are valid `type` names?

_I find pretty intuitive  to have characters like `'/'` , '.'` that would help to represent some "name scope" (like C# namespaces do) and to organize better the model definition, and it would be worth to make them "officially" supported in the doc_

**Comment by rhamzeh** [comment, marked as answer]
The correct regex is the one used by the API:
- Type: https://github.com/openfga/api/blob/main/openfga/v1/openfga.proto#L19 | `"^[^:#@\\s]{1,254}$"`
- Relation: https://github.com/openfga/api/blob/main/openfga/v1/openfga.proto#L68 | `"^[^:#@\\s]{1,50}$"`
- Condition: https://github.com/openfga/api/blob/main/openfga/v1/openfga.proto#L39C16-L39C33 | `"^[^\\s]{2,256}$"` 
- Object: https://github.com/openfga/api/blob/main/openfga/v1/openfga.proto#L80 | `"^[^\\s]{2,256}$"`

There's additionally these set of regexes defined in the API:
https://github.com/openfga/openfga/blob/main/pkg/tuple/tuple.go#L35-L38
```
	userIDRegex   = regexp.MustCompile(`^[^:#\s]+$`)
	objectRegex   = regexp.MustCompile(`^[^:#\s]+:[^#:\s]+$`)
	userSetRegex  = regexp.MustCompile(`^[^:#\s]+:[^#\s]+#[^:#\s]+$`)
	relationRegex = regexp.MustCompile(`^[^:#@\s]+$`)
```

Validation has to pass both sets of regexes in order to be valid according to the API.

---

On the VS Code/Playground side we use a different set of less permissive regexes:
https://github.com/openfga/vscode-ext/blob/07e65ccd3a0d533bcbbb2b4495899d73d2656e3b/server/src/openfga-yaml-schema.ts#L126-L154

We have yet more regexes in language:
https://github.com/openfga/language/blob/35310a2e97386126a01db1c4360e013d984f98f2/pkg/js/validator/validate-dsl.ts#L7-L10

We do plan on standardizing these, this will probably happen sometime within the next couple of months as part of a larger work to move validation ownership to the language repo.

Relevant issue: https://github.com/openfga/language/issues/226

**Reply by bubianchi-criteo** [reply]
Thank you for clarifying

**Reply by bubianchi-criteo** [reply]
Note tha the lexer of the DSL grammar should also probably updated since it defines as
 `IDENTIFIER: (LETTER | '_') (LETTER | DIGIT | '_' | MINUS)*;`
See https://github.com/openfga/language/blob/35310a2e97386126a01db1c4360e013d984f98f2/OpenFGALexer.g4#L132

**Reply by rhamzeh** [reply]
That is a different identifier - that is for the names of the types/relations, not the IDs. Yes, we will update it, though it should be more strict than the IDs.

E.g. not sure we want to allow `organ!zat!ion` as a type/relation name

**Reply by bubianchi-criteo** [reply]
I would like to propose a pull request with this kind of "extended identifier" lexeme (additionally to the original CEL grammar ) 
`EXTENDED_IDENTIFIER: (LETTER | '_')((LETTER | DIGIT | '_' | MINUS| SLASH | DOT)*(LETTER | DIGIT | '_'))*;`
that will be used in the parser rules for the names of types and relations
